kafka-client/src/test/java/dev/responsive/kafka/testutils/StoreComparator.java

```diff
@@ -0,0 +1,225 @@
+    void apply(String method, Object[] args, Object actual, Object truth);
+  }
+
+  public static class MultiKeyValueStoreSupplier implements KeyValueBytesStoreSupplier {
+    private final KeyValueBytesStoreSupplier sourceOfTruth;
+    private final KeyValueBytesStoreSupplier candidate;
+    private final CompareFunction compare;
+
+    public MultiKeyValueStoreSupplier(
+        KeyValueBytesStoreSupplier sourceOfTruth,
+        KeyValueBytesStoreSupplier candidate
+    ) {
+      this.sourceOfTruth = sourceOfTruth;
+      this.candidate = candidate;
+      this.compare = null;
+    }
+
+    public MultiKeyValueStoreSupplier(
+        KeyValueBytesStoreSupplier sourceOfTruth,
+        KeyValueBytesStoreSupplier candidate,
+        CompareFunction compare
+    ) {
+      this.sourceOfTruth = sourceOfTruth;
+      this.candidate = candidate;
+      this.compare = compare;
+    }
+
+    @Override
+    public String metricsScope() {
+      return this.sourceOfTruth.metricsScope();
+    }
+
+    @Override
+    public String name() {
+      return this.sourceOfTruth.name();
+    }
+
+    @Override
+    public KeyValueStore<Bytes, byte[]> get() {
+      if (this.compare == null) {
+        return new KeyValueStoreComparator<>(this.sourceOfTruth.get(), this.candidate.get());
+      }
+      return new KeyValueStoreComparator<>(
+          this.sourceOfTruth.get(),
+          this.candidate.get(),
+          this.compare
+      );
+    }
+  }
+
+  public static class MultiSessionStoreSupplier<K, V> implements StoreSupplier<SessionStore<K, V>> {
+    private final StoreSupplier<SessionStore<K, V>> sourceOfTruth;
+    private final StoreSupplier<SessionStore<K, V>> candidate;
+    private final CompareFunction compare;
+
+    public MultiSessionStoreSupplier(
+        StoreSupplier<SessionStore<K, V>> sourceOfTruth,
+        StoreSupplier<SessionStore<K, V>> candidate
+    ) {
+      this.sourceOfTruth = sourceOfTruth;
+      this.candidate = candidate;
+      this.compare = null;
+    }
+
+    public MultiSessionStoreSupplier(
+        StoreSupplier<SessionStore<K, V>> sourceOfTruth,
+        StoreSupplier<SessionStore<K, V>> candidate,
+        CompareFunction compare
+    ) {
+      this.sourceOfTruth = sourceOfTruth;
+      this.candidate = candidate;
+      this.compare = compare;
+    }
+
+    @Override
+    public String metricsScope() {
+      return this.sourceOfTruth.metricsScope();
+    }
+
+    @Override
+    public String name() {
+      return this.sourceOfTruth.name();
+    }
+
+    @Override
+    public SessionStore<K, V> get() {
+      if (this.compare == null) {
+        return new SessionStoreComparator<>(this.sourceOfTruth.get(), this.candidate.get());
+      }
+      return new SessionStoreComparator<>(
+          this.sourceOfTruth.get(),
+          this.candidate.get(),
+          this.compare
+      );
+    }
+  }
+
+  public static class MultiStateStoreContext implements InvocationHandler, StateRestoreCallback,
```

Comment:
nit: this feels like it should be its own class - why's it nested here alongside the factory methods?